README.md

```diff
@@ -13,6 +13,223 @@ for kubernetes bootstrap.
+
+    docker_build(controller['image'], controller['path'])
+
+```
+
+See https://github.com/chuckha/capi-dev for an example of a more complex developemt environment using [tilt](https://tilt.dev/).
+
+## How it works?
+Once your test environment is in place, create a `Cluster` object and its corresponding `DockerCluster`
+infrastructure object.
+
+```yaml
+kind: DockerCluster
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
+metadata:
+  name: my-cluster-docker
+---
+kind: Cluster
+apiVersion: cluster.x-k8s.io/v1alpha2
+metadata:
+  name: my-cluster
+spec:
+  infrastructureRef:
+    kind: DockerCluster
+    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
+    name: my-cluster-docker
+```
+
+After you can start creating machines by defining a `Cluster`, its corresponding `DockerMachine` object, and
+the `KubeadmConfig` bootstrap object.
+
+```yaml
+kind: KubeadmConfig
+apiVersion: bootstrap.cluster.x-k8s.io/v1alpha2
+metadata:
+  name: my-control-plane1-config
+spec:
+  initConfiguration:
+    nodeRegistration:
+      kubeletExtraArgs:
+        eviction-hard: nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%
+  clusterConfiguration:
+    controllerManager:
+      extraArgs:
+        enable-hostpath-provisioner: "true"
+    networking:
+      podSubnet: "10.244.0.0/16"
+      dnsDomain: ""
+      serviceSubnet: ""
+---
+kind: DockerMachine
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
+metadata:
+  name: my-control-plane1-docker
+---
+kind: Machine
+apiVersion: cluster.x-k8s.io/v1alpha2
+metadata:
+  name: my-control-plane1
+  labels:
+    cluster.x-k8s.io/cluster-name: my-cluster
+    cluster.x-k8s.io/control-plane: "true"
+    set: controlplane
+spec:
+  bootstrap:
+    configRef:
+      kind: KubeadmConfig
+      apiVersion: bootstrap.cluster.x-k8s.io/v1alpha2
+      name: my-control-plane1-config
+  infrastructureRef:
+    kind: DockerMachine
+    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
+    name: my-control-plane1-docker
+  version: "v1.14.2"
+```
+
+CAPBK main responsibility is to convert a `KubeadmConfig` bootstrap object into a cloud-init script that is
+going to turn a Machine into a Kubernetes Node using kubeadm.
+
+The cloud-init script will be saved into the `KubeadmConfig.Status.BootstrapData` and then the infrastructure provider
+(CAPD in this example) will pick up this value and proceed with the machine creation and the actual bootstrap.
+
+#### KubeadmConfig examples
+
+Bootstrap control plane node:
+```yaml
+kind: KubeadmConfig
+apiVersion: bootstrap.cluster.x-k8s.io/v1alpha2
+metadata:
+  name: my-control-plane1-config
+spec:
+  initConfiguration:
+    nodeRegistration:
+      kubeletExtraArgs:
+        eviction-hard: nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%
+  clusterConfiguration:
+    controllerManager:
+      extraArgs:
+        enable-hostpath-provisioner: "true"
+    networking:
+      podSubnet: "10.244.0.0/16"
+      dnsDomain: ""
```

Comment:
Outside of the scope of this PR -- Should we make these optional or does an empty string indicate something different than nil?

Reply:
this is probably a leftover of https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/pull/156
cc @vincepri